pypeclub/OpenPype
# Version handling

## Goals
Be able to define which version should be used as production or staging through global settings, prepare base to be able store settings by OpenPype version and be able to manage installed zip versions.

## Description
All of this logic must be primarily in igniter which cares about it before OpenPype logic starts so the easiest way is to implement and put required logic into one place and store it to `sys.modules` so OpenPype can use it too.
OpenPype will need access to multiple functionality:
- `OpenPypeVersion` class which parse version string and gives ability to sort versions
- function to get "expected version" that should be used from global settings (staging or production)
    - should return None if version is not defined there
- ability to get latest version from all versions (staging or production)
- ability to get OpenPype version inside build
    - this will never be marked as staging
- ability to get all remote versions (staging or production)
    - hints for settings UI
- ability to get all local versions (staging or production)
    - for management of local versions (delete them etc.)
- ( ? ) install/remove local version
    - for management of local versions (delete them etc.)

Not all of these features will be used in this PR but igniter should give access to them as igniter requires to create new build.

## Changes:
- moved version detection and version handling to `OpenPypeVersion` class to allow use outside of Igniter
- `OpenPypeVersion` is stored to `sys.modules` so is accessible in any OpenPype process
- added new keys into general settings `production_version` and `staging_version`
    - both can define which version of OpenPype is used
    - if their values are empty OpenPype should work as before
    - they're text inputs with validation of parsing to match `OpenPypeVersion` semantic regex
    - in UI have validation of regex and validation of existence
        - added label which tells user if the value is ok or not
        - existence check is not strict to not "break" settings if version is not available on machine where settings are set (e.g. in planned admin UI) and the check would not cover all possible issues
- the new keys `production_version` or `staging_version` are forced to be used on OpenPype start
    - it is possible to use different version with env `OPENPYPE_VERSION` or with argument `--use-version=...`
- it is possible to use latest version with `--use-version=latest`
- OpenPype won't start if version in global settings is specified and is not available for the machine
    - only way how to go around is to set `OPENPYPE_VERSION` environment or pass `--use-version=...` argument
- Qt Application is not created if already exists
- added function to show dialog with title, message and `OK` button to a user
- settings entity `text` can have `value_hints` which will show a value hints in UI
- new `OpenPypeVersion` methods
    - `get_version_string_from_directory(dirpath)` - get version as string (None if was not found)
    - `get_versions_from_directory(dirpath)` - get list of OpenPype version from directory
    - `get_local_versions(production, staging)` - get list of OpenPype versions available on user's machine
        - can return only staging, only production or both
    - `get_remote_versions(production, staging)` - get list of OpenPype versions available in version repository path
        - can return only staging, only production or both
    - `get_installed_version` - version of OpenPype inside build
    - `get_latest_version(staging, local, remote)` - get latest version including build version for non staging variant
        - looks for production or staging in local repository, remote repository or both repositories
        - returns `None` if nothing was found (can happen only if looking for staging and there are no staging zip versions)
    - `get_expected_studio_version(staging)` - get version defined by studio in settings
        - returns `None` if studio does not have defined a version in global settings
- new methods in`BootstrapRepos`
    - `find_openpype_version(version, staging)` to look for openpype version with path in specific order
        1. check if match build version
        2. check if is directory in local version
        3. check if is zip in local version
        4. check if exists in remote repository
        - returns `None` if was not found at all
    - `find_latest_openpype_version(staging)` to look for latest openpype version with path
        - use same order as `find_openpype_version` to determine which is returned
        - returns `None` if was not found any version
- show message dialog if version was not found

## Question
### Version in build
Currently version in build is never marked as staging even if has `+staging` in version.py. Should we allow that?
**Why is that important**: It is not possible to have build version as a "staging version" for studio. It is possible to set it but it won't be used.

## Plans for future
- install/remove local version
    - maybe auto remove
- it would also be nice to have UI for that available in igniter
